server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java

```diff
@@ -96,37 +95,53 @@ public DataSegment push(File dataSegmentFile, DataSegment segment, boolean repla
 
     final File tmpOutDir = new File(baseStorageDir, intermediateDirFor(storageDir));
     log.info("Creating intermediate directory[%s] for segment[%s]", tmpOutDir.toString(), segment.getIdentifier());
-    final long size = compressSegment(dataSegmentFile, tmpOutDir);
+    FileUtils.forceMkdir(tmpOutDir);
 
-    final DataSegment dataSegment = createDescriptorFile(
-        segment.withLoadSpec(makeLoadSpec(new File(outDir, "index.zip").toURI()))
-               .withSize(size)
-               .withBinaryVersion(SegmentUtils.getVersionFromDir(dataSegmentFile)),
-        tmpOutDir
-    );
+    try {
+      final File tmpIndexFile = new File(tmpOutDir, INDEX_FILENAME);
+      final long size = compressSegment(dataSegmentFile, tmpIndexFile);
 
-    // moving the temporary directory to the final destination, once success the potentially concurrent push operations
-    // will be failed and will read the descriptor.json created by current push operation directly
-    FileUtils.forceMkdir(outDir.getParentFile());
-    if (replaceExisting) {
-      try {
-        FileUtils.cleanDirectory(outDir);
-      }
-      catch (IllegalArgumentException e) {
-        log.debug("Directory [%s] doesn't exist and couldn't be cleaned, this is normal", outDir);
-      }
-      Files.move(tmpOutDir.toPath(), outDir.toPath(), StandardCopyOption.REPLACE_EXISTING);
-    } else {
-      try {
-        Files.move(tmpOutDir.toPath(), outDir.toPath());
-      }
-      catch (FileAlreadyExistsException e) {
-        log.info("Push destination directory[%s] exists, ignore this message if replication is configured.", outDir);
-        FileUtils.deleteDirectory(tmpOutDir);
-        return jsonMapper.readValue(new File(outDir, "descriptor.json"), DataSegment.class);
+      final File tmpDescriptorFile = new File(tmpOutDir, DESCRIPTOR_FILENAME);
+      DataSegment dataSegment = createDescriptorFile(
+          segment.withLoadSpec(makeLoadSpec(new File(outDir, INDEX_FILENAME).toURI()))
+                 .withSize(size)
+                 .withBinaryVersion(SegmentUtils.getVersionFromDir(dataSegmentFile)),
+          tmpDescriptorFile
+      );
+
+      FileUtils.forceMkdir(outDir);
+      if (replaceExisting) {
+        Files.move(
```

Comment:
Better to use `renameTo` on the actual file objects. `Files.move` is a convenience method that falls back to copy-then-delete when rename fails (possibly because the files are on different filesystems). But we really don't want that. It would defeat the purpose of using a temp file and then renaming it (atomicity).

In this case, it shouldn't matter, since we "know" that the files are on the same filesystem anyway. But I think it's best to encode that knowledge in our choice of function used to rename this file.

Reply:
@gianm I'm not sure how to use File.renameTo(File) to accomplish this since that method doesn't support overwriting existing files.

Reply:
@dclim `renameTo` says it might not succeed if the destination already exists, but you can trust that it will, at least on a POSIX system, which are the only kind that we support anyway.